plan/stats.go

```diff
@@ -0,0 +1,178 @@
+	}
+	return p.profile
+}
+
+func (p *Selection) statsProfile() *statsProfile {
+	if p.profile != nil {
+		return p.profile
+	}
+	childProfile := p.children[0].(LogicalPlan).statsProfile()
+	// TODO: estimate it by histogram.
+	p.profile = &statsProfile{
+		count:       childProfile.count * selectionFactor,
+		cardinality: make([]float64, len(childProfile.cardinality)),
+	}
+	for i := range p.profile.cardinality {
+		p.profile.cardinality[i] = childProfile.cardinality[i] * selectionFactor
+	}
+	return p.profile
+}
+
+func (p *Union) statsProfile() *statsProfile {
+	if p.profile != nil {
+		return p.profile
+	}
+	p.profile = &statsProfile{
+		cardinality: make([]float64, p.schema.Len()),
+	}
+	for _, child := range p.children {
+		childProfile := child.(LogicalPlan).statsProfile()
+		p.profile.count += childProfile.count
+		for i := range p.profile.cardinality {
+			p.profile.cardinality[i] += childProfile.cardinality[i]
+		}
+	}
+	return p.profile
+}
+
+func (p *Limit) statsProfile() *statsProfile {
+	if p.profile != nil {
+		return p.profile
+	}
+	childProfile := p.children[0].(LogicalPlan).statsProfile()
+	p.profile = &statsProfile{
+		count:       float64(p.Count),
+		cardinality: make([]float64, len(childProfile.cardinality)),
+	}
+	if p.profile.count > childProfile.count {
+		p.profile.count = childProfile.count
+	}
+	for i := range p.profile.cardinality {
+		p.profile.cardinality[i] = childProfile.cardinality[i]
+		if p.profile.cardinality[i] > p.profile.count {
+			p.profile.cardinality[i] = p.profile.count
+		}
+	}
+	return p.profile
+}
+
+func (p *TopN) statsProfile() *statsProfile {
+	if p.profile != nil {
+		return p.profile
+	}
+	childProfile := p.children[0].(LogicalPlan).statsProfile()
+	p.profile = &statsProfile{
+		count:       float64(p.Count),
+		cardinality: make([]float64, len(childProfile.cardinality)),
+	}
+	if p.profile.count > childProfile.count {
+		p.profile.count = childProfile.count
+	}
+	for i := range p.profile.cardinality {
+		p.profile.cardinality[i] = childProfile.cardinality[i]
+		if p.profile.cardinality[i] > p.profile.count {
+			p.profile.cardinality[i] = p.profile.count
+		}
+	}
+	return p.profile
+}
+
+func getCardinality(cols []*expression.Column, schema *expression.Schema, profile *statsProfile) float64 {
+	indices := schema.ColumnsIndices(cols)
+	if indices == nil {
+		log.Errorf("Cannot find column %s indices from schema %s", cols, schema)
+		return 0
+	}
+	var cardinality float64
+	for _, idx := range indices {
+		if cardinality < profile.cardinality[idx] {
+			// It is a very elementary estimation.
+			cardinality = profile.cardinality[idx]
+		}
+	}
+	return cardinality
+}
+
+func (p *Projection) statsProfile() *statsProfile {
+	if p.profile != nil {
+		return p.profile
+	}
+	childProfile := p.children[0].(LogicalPlan).statsProfile()
+	p.profile = &statsProfile{
+		count:       childProfile.count,
+		cardinality: make([]float64, len(childProfile.cardinality)),
```

Comment:
for `select a+1, a+1 from t`
the projection schema length is 2, its child cardinality length is 1.